### PR TITLE
Optimize the initialization of `List[ndarray]` in `cupy.array`

### LINKED ISSUE #409: cupy.asarray doesn't accept list of cupy.ndarray
Both `numpy.asarray` and `cupy.asarray` accept list of `numpy.ndarray`:

```
>>> numpy.asarray([numpy.zeros(1)])
array([[ 0.]])
>>> cupy.asarray([numpy.zeros(1)])
array([[ 0.]])
```

However, `cupy.asarray` doesn't accept list of `cupy.ndarray`:

```
>>> cupy.asarray([cupy.zeros(1)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/user/.pyenv/versions/flatten-obs/lib/python3.6/site-packages/cupy/creation/from_data.py", line 46, in asarray
    return core.array(a, dtype, False)
  File "cupy/core/core.pyx", line 1799, in cupy.core.core.array (cupy/core/core.cpp:58790)
  File "cupy/core/core.pyx", line 1824, in cupy.core.core.array (cupy/core/core.cpp:58526)
ValueError: Unsupported dtype object
```

This behavior might be consistent in some sense, but I often find it cumbersome when writing numpy/cupy agnostic code.

### PR DESCRIPTION
**Pull request by okapies**
- Implement `cupy.array(<List[cupy.ndarray]>)`
- Improve the performance of `cupy.array(<List[numpy.ndarray]>)`

Fixes #409 and #1084. Ref #1030.

### CODE REVIEW

**Comment by okapies** [comment]
A performance testing at https://github.com/cupy/cupy/pull/2081/commits/49668ab3e68d71ce0545e7339900589689854aaa (The script is [here](https://gist.github.com/okapies/a6a9d8adbbb29b680f4e78cca44e7221)).

## `--shape="(28, 28)"`
| batch-size | 10 | 100 | 1000 |
|:--|:--|:--|:--|
| numpy -> numpy | 0.153 | 1.246 | 12.679 |
| numpy -> cupy | 1.479 | 3.212 | 19.265 |
| cupy -> cupy | 2.718 | 4.770 | 23.148 |

## `--shape="(3, 224, 224)"`
| batch-size | 1 | 10 | 100 |
|:--|:--|:--|:--|
| numpy -> numpy | 0.454 | 3.913 | 190.949 |
| numpy -> cupy | 2.097 | 9.493 | - |
| cupy -> cupy | 1.504 | 2.775 | 8.908 |

**Comment by okapies** [comment]
I run repeated benchmarks for `numpy.array(<List[numpy.ndarray]>)` in a variety of data size. It seems that zero filling of allocated page generates huge overhead when copying large arrays.

screenshot_2019-03-07 numpy_array_perf

screenshot_2019-03-07 np_nested_array_125000_numactl log folded gz - speedscope

(Note that you should use `numactl --cpunodebind=0 --membind=0` in a NUMA environment to prevent the unpredictable result.)

**Comment by okapies** [comment]
Upon further investigation, dTLB misses and LLC cache misses cause the above performance degradation.

**Comment by okapies** [comment]
Specifying `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1` significantly reduces the `dTLB-(load|store)-misses` and improve the overhead for larger array. I'll use it for the benchmark of this PR to measure the baseline performance of my algorithm.

(Note that this configuration leads a fragmentation of the heap memory. Don't use it in production.)

**Comment by okapies** [comment]
A performance testing at https://github.com/cupy/cupy/pull/2081/commits/49668ab3e68d71ce0545e7339900589689854aaa (The script is [here](https://gist.github.com/okapies/a6a9d8adbbb29b680f4e78cca44e7221)).

## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

##  cupy-cuda==6.0.0b3
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `numpy.array(<List[numpy.ndarray]>)` | 0.466 | 3.642 | 133.286 | maybe CPU bootleneck on `memmove` |
| `cupy.array(<List[numpy.ndarray]>)` | 1.201 | 8.308 | 199.298 | 〃 |
| `cupy.array(<List[cupy.ndarray]>)` | n/a | n/a | n/a | - |

## b983599
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | - | - | - | maybe CPU bootleneck on `memmove` |
| `cupy.array(<List[cupy.ndarray]>)` | - | - | - | - |

**Comment by okapies** [comment]
## b212555
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.759 | 6.249 | 137.616 | 25-40% faster |

**Comment by okapies** [comment]
## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

## dd5ccbc
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.865 | 6.175 | 136.076 | 25-40% faster than the baseline |
| `cupy.array(<List[cupy.ndarray]>)` | 0.518 | 1.177 | 8.864 | - |

**Comment by okapies** [comment]
@niboshi Could you please check it? Thanks.

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        a.data.copy_from_host_async(ctypes.c_void_p(mem.ptr), nbytes)
+        pinned_memory._add_to_watch_list(stream.record(), mem)
+    else:
+        # fallback to numpy array and send it to GPU
+        a = _send_object_to_gpu(arrays, dtype, order, ndmin)
+
+    return a
+
+
+cpdef inline _alloc_pinned_memory(Py_ssize_t nbytes):
```
Comment (niboshi) [inline cupy/core/core.pyx]: Isn't it possible to make this function do the transfer as well, not only allocation?
The warning looks out of place.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        # write concatenated arrays to the pinned memory directly
+        src_cpu = numpy.frombuffer(mem, dtype, itemcount).reshape(
+            shape, order=order)
+        numpy.concatenate([e[None] for e in arrays], 0, src_cpu)
+        a = ndarray(shape, dtype=dtype, order=order)
+        a.data.copy_from_host_async(ctypes.c_void_p(mem.ptr), nbytes)
+        pinned_memory._add_to_watch_list(stream.record(), mem)
+    else:
+        # fallback to numpy array and send it to GPU
+        a = _send_object_to_gpu(arrays, dtype, order, ndmin)
```
Comment (niboshi) [inline cupy/core/core.pyx]: Is this fallback meaningful?
Doesn't it cause OOM again in `_send_object_to_gpu`?

File: cupy/core/core.pxd [hunk 3]
```diff
@@ -94,6 +94,14 @@ cdef class Indexer:
     cdef CPointer get_pointer(self)
 
 
+cdef tuple _get_concat_shape(object obj)
```
Comment (niboshi) [inline cupy/core/core.pxd]: I think all of the declarations in `.pxd` are not needed.

File: cupy/core/core.pyx [hunk 4]
```diff
@@ -1708,7 +1710,7 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
         else:
             a = src.copy(order=order).astype(dtype, copy=False)
 
-        ndim = a._shape.size()
+        ndim = len(a._shape)
```
Comment (niboshi) [inline cupy/core/core.pyx]: Is there any reason of this change? Is `len` faster?

File: cupy/core/core.pyx [hunk 5]
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+
+
+cdef tuple _get_concat_shape(object obj):
+    # Returns a tuple of the following:
+    # 1. concatenated shape if it can be converted to a single CuPy array by
+    #    just concatenating it (i.e., the object is a (nested) sequence only
+    #    which contains NumPy/CuPy array(s) with same shape and dtype).
+    #    Returns None otherwise.
+    # 2. dtype if the object is an array
+    # 3. a first element in the object
```
Comment (niboshi) [inline cupy/core/core.pyx]: How about using "item" instead of "element"?
"Element" more commonly refers to a scalar in arrays.

Comment (niboshi) [inline cupy/core/core.pyx]: It looks only its type is used.
How about returning the type instead?

File: cupy/testing/array.py [hunk 6]
```diff
@@ -128,6 +129,70 @@ def assert_array_list_equal(xlist, ylist, err_msg='', verbose=True):
...
+        buf = ctypes.create_string_buffer(a.nbytes)
+        a.data.copy_to_host(ctypes.c_void_p(ctypes.addressof(buf)), a.nbytes)
+        arr = bytearray(a.nbytes)
+        arr[:] = buf.raw
+        return arr
+    else:
+        return ctypes.string_at(a.ctypes.data, a.nbytes)
+
+
+def assert_array_exactly_equal(x, y, err_msg='', verbose=True):
```
Comment (niboshi) [inline cupy/testing/array.py]: Does this function compare garbage bytes between elements?
Is there any point to do that? How about comparing strides instead?

File: cupy/core/core.pyx [hunk 7]
```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+            order = 'C'
+    a_cpu = numpy.array(obj, dtype=dtype, copy=False, order=order,
+                        ndmin=ndmin)
+    a_dtype = a_cpu.dtype
+    if a_dtype.char not in '?bhilqBHILQefdFD':
+        raise ValueError('Unsupported dtype %s' % a_dtype)
+    cdef vector.vector[Py_ssize_t] a_shape = (<object>a_cpu).shape
+    cdef ndarray a = ndarray(a_shape, dtype=a_dtype, order=order)
+    if a_cpu.ndim == 0:
+        a.fill(a_cpu[()])
```
[the remark this answers is not in the document]
Comment (okapies) [reply cupy/core/core.pyx]: Hmm, it was introduced about two years ago in #157. I cannot judge this indexing (`()`) is redundant or not. How do you think?

File: cupy/core/core.pyx (hunk 1 above)
Comment (okapies) [reply cupy/core/core.pyx]: Now this is used by both `_send_object_to_gpu` and `_send_numpy_array_list_to_gpu` which have different transfer logics. One idea is that move the warning to the above functions. Do you have any ideas?

File: cupy/core/core.pyx (hunk 4 above)
Comment (okapies) [reply cupy/core/core.pyx]: It is because I saw that the compiler complains `a._shape` doesn't have `size()` method (maybe it fails to infer `_shape` as a tuple). I now checked again and `.size()` works... It is strange. Anyway, I'll fix it.

File: cupy/core/core.pxd (hunk 3 above)
Comment (okapies) [reply cupy/core/core.pxd]: Thanks. I'll remove them.

File: cupy/core/core.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (okapies) [reply cupy/core/core.pyx]: As an offline discussion, we found it can be replaced with `a.fill(a_cpu)` because `fill` calls `a_cpu.item()` internally.

File: cupy/core/core.pyx (hunk 2 above)
Comment (okapies) [reply cupy/core/core.pyx]: I'll fix to call `a.data.copy_from_host()` directly.

File: cupy/core/core.pyx (hunk 1 above)
Comment (okapies) [reply cupy/core/core.pyx]: I remembered that it is hard to move to other places because the message takes `e` as an argument.

Comment (okapies) [reply cupy/core/core.pyx]: Renamed to `_alloc_async_tranfer_buffer` for clarity.

**Comment by okapies** [comment]
@niboshi Could you please check it again?

File: cupy/testing/array.py [hunk 8]
```diff
@@ -128,6 +128,48 @@ def assert_array_list_equal(xlist, ylist, err_msg='', verbose=True):
             verbose=verbose)
 
 
+def assert_array_exactly_equal(x, y, err_msg='', verbose=True):
```
Comment (niboshi) [inline cupy/testing/array.py]: How about adding `strides_check` option in `assert_array_equal`, instead of creating a new function?

File: cupy/core/core.pyx [hunk 9]
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        raise ValueError('Unsupported dtype %s' % a_dtype)
+    cdef vector.vector[Py_ssize_t] a_shape = a_cpu.shape
+    cdef ndarray a = ndarray(a_shape, dtype=a_dtype, order=order)
+    if a_cpu.ndim == 0:
+        a.fill(a_cpu)
+        return a
+    cdef Py_ssize_t nbytes = a.nbytes
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = \
```
Comment (niboshi) [inline cupy/core/core.pyx]: It's better to use `(` for line continuation instead of `\`. (PEP-8)

File: cupy/core/core.pyx [hunk 10]
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        Py_ssize_t ndmin):
+    if dtype.char not in '?bhilqBHILQefdFD':
+        raise ValueError('Unsupported dtype %s' % dtype)
+
+    cdef ndarray a  # allocate it after pinned memory is secured
+    cdef size_t itemcount = internal.prod(shape)
+    cdef size_t nbytes = itemcount * get_dtype(dtype).itemsize
+
+    stream = stream_module.get_current_stream()
+    cdef pinned_memory.PinnedMemoryPointer mem = \
```
Comment (niboshi) [inline cupy/core/core.pyx]: Ditto.

File: cupy/core/core.pyx [hunk 11]
```diff
@@ -1717,51 +1719,190 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        a_cpu = numpy.array(arrays, dtype=dtype, copy=False, order=order,
+                            ndmin=ndmin)
+        a = ndarray(shape, dtype=dtype, order=order)
+        a.data.copy_from_host(
+            ctypes.c_void_p(a_cpu.__array_interface__['data'][0]), nbytes)
 
     return a
 
 
+cpdef inline _alloc_async_transfer_buffer(Py_ssize_t nbytes):
```
Comment (niboshi) [inline cupy/core/core.pyx]: `cpdef` -> `cdef`?

File: cupy/testing/array.py [hunk 12]
```diff
@@ -74,12 +74,14 @@ def assert_array_max_ulp(a, b, maxulp=1, dtype=None):
         cupy.asnumpy(a), cupy.asnumpy(b), maxulp=maxulp, dtype=dtype)
 
 
-def assert_array_equal(x, y, err_msg='', verbose=True):
+def assert_array_equal(x, y, strides_check=False, err_msg='', verbose=True):
```
Comment (niboshi) [inline cupy/testing/array.py]: The new argument should come last, because the order of arguments is a part of API.

File: cupy/testing/array.py [hunk 13]
```diff
@@ -90,6 +92,32 @@ def assert_array_equal(x, y, err_msg='', verbose=True):
...
+    if strides_check:
+        if x.strides != y.strides:
+            msg = ["Strides are not equal:"]
+            if err_msg:
+                msg = [msg[0] + ' ' + err_msg]
+            if verbose:
+                msg.append(" x: {}".format(x.strides))
+                msg.append(" y: {}".format(y.strides))
+            raise AssertionError('\n'.join(msg))
+        if x.flags.c_contiguous != y.flags.c_contiguous:
```
Comment (niboshi) [inline cupy/testing/array.py]: I think this is redundant. `flags.c_contiguous` and `flags.f_contiguous` are just reduced representations of `strides`.

File: cupy/testing/array.py [hunk 14]
```diff
@@ -90,6 +92,32 @@ def assert_array_equal(x, y, err_msg='', verbose=True):
         cupy.asnumpy(x), cupy.asnumpy(y), err_msg=err_msg,
         verbose=verbose)
 
+    if strides_check:
+        if x.strides != y.strides:
+            msg = ["Strides are not equal:"]
```
Comment (niboshi) [inline cupy/testing/array.py]: Please use single quotes for string literals.

File: cupy/testing/helper.py [hunk 15]
```diff
@@ -337,8 +337,8 @@ def check_func(x, y):
 
 
 def numpy_cupy_array_equal(err_msg='', verbose=True, name='xp',
-                           type_check=True, accept_error=False, sp_name=None,
-                           scipy_name=None):
+                           type_check=True, strides_check=False,
```
Comment (niboshi) [inline cupy/testing/helper.py]: Ditto. The order of arguments.

File: tests/cupy_tests/random_tests/test_generator.py [hunk 16]
```diff
@@ -100,7 +100,8 @@ def _generate_check_repro(self, func, seed):
         y = func()
         testing.assert_array_equal(
             x, y,
-            'Randomly generated arrays with the same seed did not match')
+            err_msg='Randomly generated arrays with the same seed did not '
```
Comment (niboshi) [inline tests/cupy_tests/random_tests/test_generator.py]: I think this change can be reverted.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by okapies** [comment]
Thanks. I'm checking regressions claimed by Jenkins.

**Comment by okapies** [comment]
Jenkins, test this please

**Comment by okapies** [comment]
@niboshi I noticed that the new implementation requires `numpy>=1.14.0` (released in Jan 2018) because it depends on third `out` argument of [`numpy.concatenate()`](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.concatenate.html#numpy.concatenate) to prevent memory allocation. Do you think it is OK or not?

**Comment by okapies** [comment]
Jenkins, test this please

**Comment by okapies** [comment]
## Parameters
- xp: `numpy` or `cupy`
- shape per item: `(3, 224, 224)`
- batch-size: `1`, `10` or `100`
- number: `10000`
- environment: `MALLOC_MMAP_MAX_=0 MALLOC_TRIM_THRESHOLD_=-1 numactl --cpunodebind=0 --membind=0`

## 756c54a
| batch-size | 1 | 10 | 100 | Note |
|:--|:--|:--|:--|:--|
| `cupy.array(<List[numpy.ndarray]>)` | 0.949 | 6.355 | 136.538 | 25-32% faster than the baseline |
| `cupy.array(<List[cupy.ndarray]>)` | 0.550 | 1.184 | 8.862 | - |

**Comment by okapies** [comment]
Jenkins, test this please

**Comment by okapies** [comment]
Jenkins, test this please.

**Comment by okapies** [comment]
@niboshi I added several tests about the handling of scalar value and dtype, and fixed the issues in 956eb2f...5f8af60 (Jenkins failures are the precision problem). Could you please check it and merge if you don't find any other problems?

**Comment by okapies** [comment]
Jenkins, test this please.

File: cupy/core/core.pyx [hunk 17]
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
     elif hasattr(obj, '__cuda_array_interface__'):
         return array(_convert_object_with_cuda_array_interface(obj),
                      dtype, copy, order, subok, ndmin)
+    else:  # obj is sequence, numpy array, scalar or the other type of object
+        shape, elem_type, elem_dtype = _get_concat_shape(obj)
+        if shape is not None and shape[-1] != 0:
```
Comment (niboshi) [inline cupy/core/core.pyx]: Doesn't it cause index error if shape is `()`?

File: cupy/core/core.pyx [hunk 18]
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+                # obj is Seq[cupy.ndarray]
+                lst = _flatten_list(obj)
+                if len(shape) == 1:
+                    # convert each scalar (0-dim) ndarray to 1-dim
+                    lst = [cupy.expand_dims(x, 0) for x in lst]
+
+                a = (_manipulation.concatenate_method(lst, 0)
+                                  .reshape(shape)
+                                  .astype(dtype, order=order, copy=False))
+            else:  # should not be reached here
```
Comment (niboshi) [inline cupy/core/core.pyx]: I think it **can** reach here depending on the input.
Please remove this comment.

File: cupy/core/core.pyx [hunk 19]
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+                raise ValueError(
+                    'The elements of obj are unsupported type(s)')
```
Comment (niboshi) [inline cupy/core/core.pyx]: Wording suggestion:
```suggestion
                    'Elements of obj have unsupported type(s)')
```

File: cupy/core/core.pyx [hunk 20]
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        else:
+            # obj is:
+            # - numpy array
+            # - scalar or sequence of scalar
+            # - empty sequence or sequence with elements whose shapes or
+            #   dtypes are unmatched
+            # - the other type of object
```
Comment (niboshi) [inline cupy/core/core.pyx]: A nitpick but to make it clearer:
```suggestion
            # - other types
```

File: cupy/core/core.pyx [hunk 21]
```diff
@@ -1717,51 +1719,213 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
...
+        a_cpu = numpy.array(arrays, dtype=a_dtype, copy=False, order=order,
+                            ndmin=ndmin)
+        a = ndarray(shape, dtype=a_dtype, order=order)
+        a.data.copy_from_host(
+            ctypes.c_void_p(a_cpu.__array_interface__['data'][0]), nbytes)
 
     return a
 
 
+cdef bint _is_outable_numpy_concatenate = (
```
Comment (niboshi) [inline cupy/core/core.pyx]: How about `_numpy_concatenate_has_out_argument`?

File: cupy/testing/array.py [hunk 22]
```diff
@@ -74,12 +74,14 @@ def assert_array_max_ulp(a, b, maxulp=1, dtype=None):
         cupy.asnumpy(a), cupy.asnumpy(b), maxulp=maxulp, dtype=dtype)
 
 
-def assert_array_equal(x, y, err_msg='', verbose=True):
+def assert_array_equal(x, y, err_msg='', verbose=True, strides_check=False):
```
Comment (niboshi) [inline cupy/testing/array.py]: Could you separate this change in another PR?

File: cupy/testing/helper.py [hunk 23]
```diff
@@ -624,11 +626,14 @@ def _make_all_dtypes(no_float16, no_bool, no_complex):
...
 
+    if not no_char_code:
+        dtypes += tuple([numpy.dtype(t).char for t in dtypes])
+
     return dtypes
 
 
 def for_all_dtypes(name='dtype', no_float16=False, no_bool=False,
-                   no_complex=False):
+                   no_complex=False, no_char_code=True):
```
Comment (niboshi) [inline cupy/testing/helper.py]: Ditto. Please separate this change.
We've recently decided to make separate PRs for separate topics as much as possible.

File: cupy/testing/array.py (hunk 22 above)
Comment (okapies) [reply cupy/testing/array.py]: Separated these changes to #2150.

File: cupy/testing/helper.py (hunk 23 above)
Comment (okapies) [reply cupy/testing/helper.py]: Separated these changes to #2152.

File: cupy/core/core.pyx (hunk 17 above)
Comment (okapies) [reply cupy/core/core.pyx]: `shape` is always non-empty tuple because a scalar value is *not* concatenatable and `_get_concat_shape()` always return `None` instead of `()`. I added `test_array_from_numpy_scalar` in fe85a30 to check this behavior:

```python
>>> a = np.array(0, dtype='b')
>>> a.ndim
0
>>> a.shape
()
>>> cupy.array(a)
array(0, dtype=int8)
>>> cupy.array([a])
array([0], dtype=int8)
```

File: cupy/core/core.pyx (hunk 18 above)
Comment (okapies) [reply cupy/core/core.pyx]: `isinstance(elem_type, (numpy.ndarray, cupy.ndarray))` is always true when `shape != None`. If the user sees this exception, it must be a bug of `_get_concat_shape()`.

Comment (niboshi) [reply cupy/core/core.pyx]: Then, please use `assert` instead of exception.

**Comment by okapies** [comment]
@niboshi Hi, could you please check it again?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM
I'm sorry for taking time!
